Use generics for conditions to match all types derived from string.

mneverov opened the pull request:
#### What type of PR is this?

/kind feature

#### What this PR does / why we need it:

Currently, methods related to setting/removing/finding conditions operate with condition type string, whereas most condition _types_ are declared as a custom type assignable to string ([PVC](https://github.com/kubernetes/kubernetes/blob/master/staging/src/k8s.io/api/core/v1/types.go#L612), [Pod](https://github.com/kubernetes/kubernetes/blob/master/staging/src/k8s.io/api/core/v1/types.go#L3205), [Node](https://github.com/kubernetes/kubernetes/blob/master/staging/src/k8s.io/api/core/v1/types.go#L6222) etc). This leads to constant casting to string on the call site ([1](https://github.com/k8up-io/k8up/blob/7713061e17d84f541ad18ed0fcb23f8c2e140fb0/api/v1/status.go#L30), [2](https://github.com/smartxworks/virtink/blob/318e476909cb479a61e7a0ae31f00695a1009ea4/pkg/controller/vmm_webhook.go#L112), [3](https://github.com/aws/aws-application-networking-k8s/blob/cde77632fea38e46055b9f19291db38b82be25dd/pkg/k8s/policyhelper/policy.go#L243), [4](https://github.com/grafana/tempo-operator/blob/c39140e6c2b2351ce1baffb88c1cf06a6ad93445/internal/status/monolithic.go#L100), [5](https://github.com/openshift-kni/lifecycle-agent/blob/40315766ef501b72a75a9944cba5b9e853ae6c5c/controllers/ibu_controller.go#L252) etc):

```go
RemoveStatusCondition(&conditions, string(corev1.PodReady))
```

This patch introduces a type constraint `~string` so casting on the call site is unnecessary (see the added tests).

#### Special notes for your reviewer:

#### Does this PR introduce a user-facing change?

```release-note
If a condition type is defined as a type assignable to string, then casting condition type to string is unnecessary when calling meta.*StatusCondition functions.

```

mneverov committed:
Use generics for conditions to match all types derived from string.

mneverov commented:
/sig api-machinery

mneverov commented:
/retest

cici37 commented:
/assign @aaron-prindle Thanks:)
/triage accepted

aaron-prindle commented:
We have to be careful here as the changes here are part of the `staging/*` directory which has code that are part of exported go libraries that other users consume.  As such we need to be sure of the backwards compatibility of the change here.

mneverov commented:
> ... code that are part of exported go libraries that other users consume

that is the idea. Users will not need to cast their custom conditions to string any more.

> .. backwards compatibility of the change

agree. Please take a look at the test. Passing a string to the functions still work, but also a custom type I added.

MikeSpreitzer commented:
Would this PR enable those utilities to be usable with Pods? I see a custom condition _struct_ for pods (https://github.com/kubernetes/kubernetes/blob/v1.31.1/staging/src/k8s.io/api/core/v1/types.go#L3250-L3270), so utilities that deal in `metav1.Condition` will not work with `PodCondition`, right?

mneverov commented:
> Would this PR enable those utilities to be usable with Pods? I see a custom condition _struct_ for pods (https://github.com/kubernetes/kubernetes/blob/v1.31.1/staging/src/k8s.io/api/core/v1/types.go#L3250-L3270), so utilities that deal in `metav1.Condition` will not work with `PodCondition`, right?

Updated the PR description -- /s/most condition/most condition _types_.

Methods I changed accept a condition type as a string that corresponds to [metav1.Condition.Type](https://github.com/kubernetes/kubernetes/blob/v1.31.1/staging/src/k8s.io/apimachinery/pkg/apis/meta/v1/types.go#L1569). In reality, most condition types are declared as a custom type that is assignable to string like [PodConditionType](https://github.com/kubernetes/kubernetes/blob/v1.31.1/staging/src/k8s.io/api/core/v1/types.go#L3205). In order to use `meta.FindStatusCondtition` method one need to cast and the method call looks like `FindStatusCondtition(&conditions, string(corev1.PodReady))`. 
After this change is merged you can call the meta functions without casting i.e. `FindStatusCondtition(&conditions, corev1.PodReady)`. 
The current behavior with the strings still work (see the test).

mneverov commented:
Also, see the [api conventions](https://github.com/kubernetes/community/blob/master/contributors/devel/sig-architecture/api-conventions.md#primitive-types):

> Do not use enums. Use aliases for string instead (e.g. NodeConditionType).

aojea commented:
I see, so you are using the Type Inference property to keep the same signature https://go.dev/blog/intro-generics#type-inference

Can we guarantee the type inference will apply always and nobody has to modify their function signatures after this?
Seeing the examples, you still need to cast to string in the other places, ...

mneverov commented:
> I see, so you are using the Type Inference property to keep the same signature https://go.dev/blog/intro-generics#type-inference
> 
> Can we guarantee the type inference will apply always and nobody has to modify their function signatures after this? 

From the [Go language specification](https://go.dev/ref/spec#General_interfaces):

> The type set of a term of the form ~T is the set of all types whose underlying type is T.

also from [the specification](https://go.dev/ref/spec#Satisfying_a_type_constraint):

> A type T satisfies a constraint C if
> 
> T [implements](https://go.dev/ref/spec#Implementing_an_interface) C; or ...

And the [Go blog](https://go.dev/blog/intro-generics#type-inference):

>  The expression ~string means the set of all types whose underlying type is string. **This includes the type string itself as well as all types declared with definitions such as type MyString string**

Please check example in [playground](https://go.dev/play/p/shYDNJIn83z) (similar to what is happening in the test).


> Seeing the examples, you still need to cast to string in the other places, ...

Existing code that uses the meta package is not affected by this change. New code can be written without casting to string. 

The gimmick here is that we still need to cast to string, since `ConditionType` is defined in `metav1` as string. But we do that inside the functions once, instead forcing the API consumers to do that. The operation `condTypeStr := string(conditionType)` is safe as described above (type ~string has underlying type string and casting string(string) is also fine).

mneverov commented:
[According to git](https://github.com/search?q=%2Fmeta%5C..*StatusCondition.*tring%2F+language%3AGo&type=code&ref=advsearch) there are ~1.5K examples of using meta functions with casting condition types to string.

mneverov commented:
@aojea @aaron-prindle @MikeSpreitzer do you need more info for the review?

aojea commented:
> @aojea @aaron-prindle @MikeSpreitzer do you need more info for the review?

I don't see a problem the way it is explained, but this is under apimachinery,you need apimachinery leads to sign off

/assign @deads2k @jpbetz

aaron-prindle commented:
> @aojea @aaron-prindle @MikeSpreitzer do you need more info for the review?

+1, I don't see an issue given the explanation above. 

/lgtm

aojea commented:
/test pull-kubernetes-apidiff

aojea commented:
/lgtm cancel

it seems this is detected as an incompatible change
```
## ./staging/src/k8s.io/apimachinery
Incompatible changes:
- ./pkg/api/meta.FindStatusCondition: changed from func([]k8s.io/apimachinery/pkg/apis/meta/v1.Condition, string) *k8s.io/apimachinery/pkg/apis/meta/v1.Condition to func([]k8s.io/apimachinery/pkg/apis/meta/v1.Condition, T) *k8s.io/apimachinery/pkg/apis/meta/v1.Condition
- ./pkg/api/meta.IsStatusConditionFalse: changed from func([]k8s.io/apimachinery/pkg/apis/meta/v1.Condition, string) bool to func([]k8s.io/apimachinery/pkg/apis/meta/v1.Condition, T) bool
- ./pkg/api/meta.IsStatusConditionPresentAndEqual: changed from func([]k8s.io/apimachinery/pkg/apis/meta/v1.Condition, string, k8s.io/apimachinery/pkg/apis/meta/v1.ConditionStatus) bool to func([]k8s.io/apimachinery/pkg/apis/meta/v1.Condition, T, k8s.io/apimachinery/pkg/apis/meta/v1.ConditionStatus) bool
- ./pkg/api/meta.IsStatusConditionTrue: changed from func([]k8s.io/apimachinery/pkg/apis/meta/v1.Condition, string) bool to func([]k8s.io/apimachinery/pkg/apis/meta/v1.Condition, T) bool
- ./pkg/api/meta.RemoveStatusCondition: changed from func(*[]k8s.io/apimachinery/pkg/apis/meta/v1.Condition, string) bool to func(*[]k8s.io/apimachinery/pkg/apis/meta/v1.Condition, T) bool
```

mneverov commented:
It appears it is not compatible change as mentioned [here](https://go-review.googlesource.com/c/tools/+/618215/comment/13e96830_e726ca79/). Example https://go.dev/play/p/iL1qVdPAGDZ

aojea commented:
> See [golang/go#67051](https://github.com/golang/go/issues/67051), also similar change in [core](https://github.com/kubernetes/kubernetes/pull/120592).

yeah, the similar change was very constrained to the repo, these function are widely used as you showed in a previous comment, but be safe than sorry, is just typing a few more characters anyway

aojea commented:
should we close @mneverov ?

mneverov commented:
Closing as the change not backward compatible
